File: projects/07/clojure/src/array_processing.clj
;; array_processing.clj "test-drives" function read_array() and print_array().
;;
;; Output: prints out array
;;
;; Usage: clojure -m array_processing
;;
;; Completed by: Justin Voss
;; Date: 2/24/22
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(ns array_processing)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; read_array() reads array values
;; Receive: aVec, a vector of numbers.
;; Return: the array, filled out with user inputted values
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn read_array [size aVec]
    (println "Enter array value: ")
  (if (= size 0)                                          ; if size is less than 0, the recursion ends, and the program returns the vector
    aVec
        (do
                (recur (- size 1) (conj  aVec (read))  )  ; recurs with the size minus one, to eventually get out of the loop when size = 0. 
        )                                                 ; the vec that is recurred is a list with a value from the user added each time
  )
)


;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; print_array() prints out an array
;; Receive: aVec, a vector of numbers. 
;; Return: prints out the vector
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn print_array [aVec]
    (if (not-empty aVec)            ; if the vector is not empty, then it prints and recurs
        (do 
            (println (first aVec))  ; prints the first value of the vector
            (recur (rest aVec))     ; recurs the rest of the list except the first value
        )
    )
)  


;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; main function to test functions read_array() and print_array()
;; Output: the sum and average of some test vectors.
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn -main []
    (println "Enter size of array: ")
  (let
    [
        vec0 []                         ; makes an empty vector
        size (read)                     ; gets a size from the user
        vector1 (read_array size vec0)  ; adds the values to vector1, to be used later to print out the vector. The read_array function returns a vector
    ]
    (println "Array: ")
    (print_array vector1)               ; prints out the vector

    
  )
)

File: projects/07/array_processing.rb
# array_processing.rb prints out an array that was inputted by the user
# Precondition: theArray is an array of numbers
# Output: the average of the values in theArray
#
# Begun by: Dr. Adams, for CS 214 at Calvin College.
# Completed by:: Justin Voss
# Date:   2/24/22
########################################################

###############################################
# read_array() reads an array from the user
# Receive: anArray, an array of numbers, and size
# Return: the array, with filled in values
################################################

def read_array(anArray,size)
    for i in 0..size-1
        anArray[i] = gets.to_i  # gets values from the user
    end

 end
 
 
 ###############################################
 # print_array() prints out the array
 # Receive: anArray, an array of numbers
 # Return: prints out the array
 ################################################
 def print_array(anArray,size)
    for i in 0..size-1 
       puts anArray[i]  # prints the array
    end
  
 
 end
 
 if __FILE__ == $0
   
    puts "Enter array size: "
    size = gets.to_i
    array1 = Array.new(size)  # Define array1 as an Array with the size as the user inputted size
    puts "Enter array values: "
    read_array(array1,size)  # calls the read array function
    puts "Array: "
    print_array(array1,size) # calls the print array function
   
 
 end
 
